File: packages/evals/tasks/bench/extract/extract_regulations_table.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";
import { normalizeString } from "../../../framework/textScoring.js";

export default defineBenchTask(
  { name: "extract_regulations_table" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto(
        "https://browserbase.github.io/stagehand-eval-sites/sites/ncc-numbering-plan/",
      );

      const tableSelector = "#field_csv-0-csvfiletable";

      const { data: allottees } = await stagehand.extract(
        "Extract ALL of the allottees and their corresponding name, area, area code, and access code.",
        z.object({
          allottee_list: z.array(
            z.object({
              allottee_name: z.string(),
              area: z.string(),
              area_code: z.string(),
              access_code: z.string(),
            }),
          ),
        }),
        { selector: tableSelector },
      );

      const allotteeList = allottees.allottee_list;
      const expectedAllottees = await page.evaluate((selector) => {
        const table = document.querySelector(selector);
        if (!table) throw new Error(`Expected regulations table not found: ${selector}`);
        return Array.from(table.querySelectorAll("tbody > tr")).flatMap((row) => {
          const cells = Array.from(row.querySelectorAll(":scope > td")).map(
            (cell) =>
              cell.querySelector(".tablesaw-cell-content")?.textContent?.trim() ??
              cell.textContent?.trim() ??
              "",
          );
          if (cells.length < 5) return [];
          return [
            {
              allottee_name: cells[1],
              area: cells[2],
              area_code: cells[3],
              access_code: cells[4],
            },
          ];
        });
      }, tableSelector);
      if (expectedAllottees.length === 0) {
        throw new Error("Expected regulations table contained no allottee rows");
      }
      const key = (item: {
        allottee_name: string;
        area: string;
        area_code: string;
        access_code: string;
      }) =>
        [
          normalizeString(item.allottee_name),
          normalizeString(item.area),
          item.area_code.trim(),
          item.access_code.trim(),
        ].join("|");
      const actualKeys = allotteeList.map(key).sort();
      const expectedKeys = expectedAllottees.map(key).sort();
      const isRegulationsCorrect =
        actualKeys.length === expectedKeys.length &&
        JSON.stringify(actualKeys) === JSON.stringify(expectedKeys);

      return {
        _success: isRegulationsCorrect,
        regulationsData: allottees,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/extract_regulations_table.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/extract_regulations_table.ts

Evals.define_task("extract_regulations_table") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/ncc-numbering-plan/")

  # The locator engine prefix is required for XPath selectors.
  locator = t.page.locator(
    "xpath=/html/body/div[3]/main/div[2]/div[2]/div/div/div[2]/article/div[2]/div[1]/div/table",
  )

  result = t.stagehand.extract(
    "Extract ALL of the Allottees and their corresponding name, area, and area code.",
    schema: {
      "type" => "object",
      "properties" => {
        "allottee_list" => {
          "type" => "array",
          "items" => {
            "type" => "object",
            "properties" => {
              "allottee_name" => { "type" => "string" },
              "area" => { "type" => "string" },
              "area_code" => { "type" => "string" },
              "access_code" => { "type" => "string" },
            },
            "required" => %w[allottee_name area area_code access_code],
            "additionalProperties" => false,
          },
        },
      },
      "required" => ["allottee_list"],
      "additionalProperties" => false,
    },
    locator: locator,
  )

  allottees = result.data

  allottees_expected_first = {
    "allottee_name" => "101 Communications Limited",
    "area" => "Lagos",
    "area_code" => "0201",
    "access_code" => "249",
  }

  allottees_expected_last = {
    "allottee_name" => "Airtel Networks Limited",
    "area" => "National",
    "area_code" => "0708",
    "access_code" => "708",
  }

  expected_length = 25

  allottee_list = allottees["allottee_list"]

  # Check that the first entry, last entry, and total number match expectations
  is_first_correct = allottee_list[0] == allottees_expected_first
  is_last_correct = allottee_list[-1] == allottees_expected_last
  is_length_correct = allottee_list.length == expected_length

  is_regulations_correct = is_first_correct && is_last_correct && is_length_correct

  { _success: is_regulations_correct, regulationsData: allottees }
end
